### PR TITLE
Support templates in serviceCluster configuration

### PR DESCRIPTION
**Pull request by douglas-reid**
This PR was created in response to https://github.com/istio/istio/issues/36162. It attempts to allow user control over `serviceCluster` names, based on golang templates over `model.BootstrapNodeMetadata`. This would avoid having to create per-workload-instance annotations across an installation.

WARNING: This would be exposing, in some fashion, Istio implementation details as part of a "hidden" API.

With this PR, the user/admin would install with config like:

```yaml
defaultConfig:
  proxyConfig:
    serviceCluster: {{ .Labels.app }}
```

to customize the `serviceCluster` behavior for an entire mesh.

- [ X ] Policies and Telemetry
- [ X ] User Experience

### CODE REVIEW

File: pkg/bootstrap/config.go [hunk 1]
```diff
@@ -278,9 +280,20 @@ func getServiceCluster(metadata *model.BootstrapNodeMetadata) string {
...
+	tmpl, err := template.New("serviceCluster").Parse(serviceCluster)
+	if err != nil {
+		log.Infof("could not parse template: %v", err)
+		return serviceCluster
+	}
+	buf := &bytes.Buffer{}
+	err = tmpl.Execute(buf, metadata)
+	if err != nil {
+		log.Warnf("could not execute template: %v", err)
+		return serviceCluster
```
Comment (kyessenov) [inline pkg/bootstrap/config.go:294]: Returning the template does not sound right.

**Comment by kyessenov** [comment]
I think this is really an API change. While I'm generally OK with this, I think it would be better to have a list of pre-determined and reasonable options.

**Comment by douglas-reid** [comment]
> I think this is really an API change. While I'm generally OK with this, I think it would be better to have a list of pre-determined and reasonable options.

I'm fairly conflicted about this myself.

How do you propose to expose the options?

File: pkg/bootstrap/config.go (hunk 1 above)
Comment (douglas-reid) [reply pkg/bootstrap/config.go]: fixed.

**Comment by douglas-reid** [comment]
@kyessenov @Stono @howardjohn wdyt about something like: https://github.com/istio/api/compare/master...douglas-reid:service-cluster-config?expand=1 as an alternative?

**Comment by Stono** [comment]
I'm happy with either, the latter puts more emphasis on istio to ensure each of those enums remains functional across releases however whereas the former puts more control in the users hands at the risk of future breaking 🤷 

In my personal case the latter would be preferred as it's then almost a guaranteed API.  It could always be revisited with an additional `TEMPLATE` option or something in the future, should the requirement for more flexibility arise?

**Pull request closed without merging**